tests/test_cleanup.py

```diff
@@ -487,3 +507,810 @@ def test_file_vdi_delete(self, mock_lock):
+                                             cleanup.VDI.DB_COALESCE:
+                                             "true"}}
+
+        res = sr.findLeafCoalesceable()
+        self.assertEqual(res, [])
+        mock_log.assert_called_with("Leaf-coalesce disabled for this SR")
+
+    @mock.patch('cleanup.Util.log')
+    def test_findLeafCoalesceable_forbiden5(self, mock_log):
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        sr.xapi.srRecord = {"other_config": {cleanup.VDI.DB_LEAFCLSC:
+                                             cleanup.VDI.LEAFCLSC_FORCE,
+                                             cleanup.VDI.DB_COALESCE:
+                                             "false"}}
+
+        res = sr.findLeafCoalesceable()
+        self.assertEqual(res, [])
+        mock_log.assert_called_with("Coalesce disabled for this SR")
+
+# Utils for testing gatherCandidates.
+
+    def srWithOneGoodVDI(self, mock_getConfig, goodConfig):
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+
+        vdi_uuid = uuid4()
+        if goodConfig:
+            mock_getConfig.side_effect = goodConfig
+        else:
+            mock_getConfig.side_effect = iter(["good", False, "blah", "blah"])
+        good = cleanup.VDI(sr, str(vdi_uuid), False)
+        sr.vdis = {"good": good}
+        return sr, good
+
+    def addBadVDITOSR(self, sr, config, coalesceable=True):
+        vdi_uuid = uuid4()
+        bad = cleanup.VDI(sr, str(vdi_uuid), False)
+        bad.getConfig = mock.MagicMock(side_effect=iter(config))
+        bad.isLeafCoalesceable = mock.MagicMock(return_value=coalesceable)
+        sr.vdis.update({"bad": bad})
+        return bad
+
+    def gather_candidates(self, mock_getConfig, config, coalesceable=True,
+                          failed=False, expected=None, goodConfig=None):
+        sr, good = self.srWithOneGoodVDI(mock_getConfig, goodConfig)
+        bad = self.addBadVDITOSR(sr, config, coalesceable=coalesceable)
+        if failed:
+            sr._failedCoalesceTargets = [bad]
+
+        res = []
+        sr.gatherCandidates(res)
+        self.assertEqual(res, [good])
+
+    @mock.patch("cleanup.AUTO_ONLINE_LEAF_COALESCE_ENABLED", True)
+    @mock.patch('cleanup.SR.leafCoalesceForbidden', autospec=True,
+                return_value=False)
+    @mock.patch('cleanup.VDI.isLeafCoalesceable', autospec=True,
+                return_value=True)
+    @mock.patch('cleanup.VDI.getConfig', autospec=True)
+    def test_gather_candidates_leaf_not_coalescable(self, mock_getConfig,
+                                                    mock_isLeafCoalesceable,
+                                                    mock_leafCoalesceForbidden
+                                                    ):
+
+        """ The bad vdi returns false for isLeafCoalesceable and is not
+            added to the list.
+        """
+        self.gather_candidates(mock_getConfig,
```

Comment:
nothing to check in these tests? How do we know they've done the right thing?

Reply:
I refactored the actual asserts into gather_candidates, The asserts definitely fire in case of an error.

Perhaps there is something reason not to put the assert in sub function, rather than a test_function??